core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/QueryEvaluationContext.java

```diff
@@ -31,28 +34,64 @@
 public interface QueryEvaluationContext {
 
 	public class Minimal implements QueryEvaluationContext {
+		// Doubly checked locking is simpler but with Project Loom coming
+		// synchronized on this is less interesting.
+		private static final VarHandle HANDLE_TO_THE_NOW_LITERAL;
+		static {
+			try {
+				HANDLE_TO_THE_NOW_LITERAL = MethodHandles
+						.privateLookupIn(QueryEvaluationContext.Minimal.class, MethodHandles.lookup())
+						.findVarHandle(QueryEvaluationContext.Minimal.class, "nowInLiteral", Literal.class);
+
+			} catch (NoSuchFieldException | IllegalAccessException e) {
+				throw new UnsupportedOperationException("The nowInLiteral field is missing");
+			}
+		}
+		// The now time as a literal, lazy set if not yet known.
+		// No need for volatile as we use the varhandle for this.
+		private Literal nowInLiteral;
+		private final Dataset dataset;
+		private final ValueFactory vf;
+
+		/**
+		 * Set the shared now value to an prexisting object
+		 * 
+		 * @param now     that is shared.
+		 * @param dataset that a query should use to the evaluate
+		 */
 		public Minimal(Literal now, Dataset dataset) {
 			super();
-			this.now = now;
+			this.nowInLiteral = now;
 			this.dataset = dataset;
+			this.vf = SimpleValueFactory.getInstance();
+			;
 		}
 
+		/**
+		 * @param dataset that a query should use to the evaluate
+		 */
 		public Minimal(Dataset dataset) {
 			this.dataset = dataset;
+			this.vf = SimpleValueFactory.getInstance();
 		}
 
-		private Literal now;
-		private final Dataset dataset;
+		/**
+		 * @param dataset that a query should use to the evaluate
+		 */
+		public Minimal(Dataset dataset, ValueFactory vf) {
+			this.dataset = dataset;
+			this.vf = vf;
+		}
 
 		@Override
 		public Literal getNow() {
 			// creating a new date is expensive because it uses the XMLGregorianCalendar implementation which is very
-			// complex
-			if (now == null) {
-				now = SimpleValueFactory.getInstance().createLiteral(new Date());
+			// complex.
+			if (nowInLiteral == null) {
+				nowInLiteral = (Literal) HANDLE_TO_THE_NOW_LITERAL.compareAndExchange(null,
+						vf.createLiteral(new Date()));
```

Comment:
<img width="1231" alt="Screenshot 2022-04-22 at 09 46 17" src="https://user-images.githubusercontent.com/797185/164641127-2e551b7c-02d2-4564-97cb-c2b59679ee74.png">


Also the javadocs for `compareAndExchange` say that the expected value is returned, if I understand them correctly: https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/invoke/VarHandle.html#compareAndExchange(java.lang.Object...)